### PR TITLE
Map Markers viewmodel placeholder

### PR DESCRIPTION
**Pull request by TheTexanCodeur**
This pull request implements a map pin viewmodel that enables the developer to add markers on the google map. Here are the main features brought with this PR:

1. **Map pin class** : This new custom map pin class will be used to represent our markers on the map. A map pin has an ID, a position and a callback function (Can be used to navigate to a post for example).

2. **Map pin provider** : This new viewmodel is used by the map UI to retrieve the markers that we want to show on the map. The UI handles the translation from our custom pin class to the google "Marker" class.

3. **Testing** : additional tests are implemented for this new feature. They verify that the new MapPin class implements hash/== correctly and that the viewmodel works with mock data (translation to "Markers").

Acceptance criteria:
- [x] Provide the view with a custom map pins provider
- [x] Create a custom map pin class that devs can use to create pins (position, tag, callback function)

### CODE REVIEW

**Review by yoannLafore — COMMENTED** [review]
Thank you for your work !
This PR will bring us one step closer to displaying pins on the map which is a key feature of our app.

Overall, I think that you did a good job with the generic approach for the map pin :+1: 

I have just made a some suggestions and questions below.

Also, given that the real implementation of the viewmodel is not done in this PR, I think that we may want to rename it to something like **Map Markers Viewmodel Placeholder**.

Other than that, thank for your contribution !
Good job :tada:

File: lib/utils/ui/map_pin.dart
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply lib/utils/ui/map_pin.dart:1]: Fixed with 7a6849b

File: lib/utils/ui/map_pin.dart [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import "package:google_maps_flutter/google_maps_flutter.dart";
+
+//custom map pin class used to display pins on the map
+class MapPin {
+  const MapPin({
+    required this.id,
+    required this.position,
+    required this.callbackFunction,
```
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply lib/utils/ui/map_pin.dart]: Fixed with 10ea364

File: lib/utils/ui/map_pin.dart [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  });
+
+  final String id;
+  final LatLng position;
+  final void Function()? callbackFunction;
```
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply lib/utils/ui/map_pin.dart]: I thought that in order to give more modularity to the developer, this attribute should be nullable. There might be cases where we don't want to trigger any action when we click on a pin.

File: test/utils/ui/map_pin_test.dart
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply test/utils/ui/map_pin_test.dart:1]: Fixed with 9376d90

File: lib/utils/ui/map_pin.dart [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other)) return true;
```
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply lib/utils/ui/map_pin.dart]: Fixed with 15b4e79

File: lib/views/home_content/map/post_map.dart [hunk 4]
```diff
@@ -33,6 +39,20 @@ class PostMap extends ConsumerWidget {
...
+    //translates the list of map pins into markers
+    for (final pin in mapPins) {
+      markers.add(
+        Marker(
+          markerId: MarkerId(pin.id),
+          position: pin.position,
+          onTap: pin.callbackFunction,
+        ),
+      );
+    }
```
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply lib/views/home_content/map/post_map.dart]: Fixed with [d9c9e7d](https://github.com/ProximaEPFL/proxima/pull/231/commits/d9c9e7d71a9e4d235278d709b2bf5701c575cd66)

File: test/utils/ui/map_pin_test.dart [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      );
+
+      final actualHash = mapPin.hashCode;
+
+      expect(actualHash, expectedHash);
+    });
+
+    test("equality overrides correctly", () {
+      //both pin need to have the same method, not just two different methods that have the same logic
+      testFunction() {}
```
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply test/utils/ui/map_pin_test.dart]: Fixed with [2784eb8](https://github.com/ProximaEPFL/proxima/pull/231/commits/2784eb8120f1e98322705d813a6c17ac9b366a2c)

File: test/utils/ui/map_pin_test.dart [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+      MapPin mapPin = MapPin(
+        id: "1",
+        position: const LatLng(0, 0),
+        callbackFunction: testFunction(),
```
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply test/utils/ui/map_pin_test.dart]: FIxed with [2784eb8](https://github.com/ProximaEPFL/proxima/pull/231/commits/2784eb8120f1e98322705d813a6c17ac9b366a2c)

File: test/utils/ui/map_pin_test.dart [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    test("inequality test on id", () {
+      testFunction() {}
+
+      MapPin mapPin = MapPin(
+        id: "1",
+        position: const LatLng(0, 0),
+        callbackFunction: testFunction(),
+      );
+
+      MapPin mapPinCopy = MapPin(
```
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply test/utils/ui/map_pin_test.dart]: Fixed with [2784eb8](https://github.com/ProximaEPFL/proxima/pull/231/commits/2784eb8120f1e98322705d813a6c17ac9b366a2c)

File: test/views/pages/map/static_map_test.dart [hunk 8]
```diff
@@ -80,4 +84,64 @@ void main() {
...
+      when(
+        mockGeolocator.getPositionStream(
+          locationSettings: geoLocationService.locationSettings,
+        ),
+      ).thenAnswer(
+        (_) => Stream.fromIterable([
+          getSimplePosition(latitude, longitude),
+          getSimplePosition(latitude + 1, longitude),
+        ]),
+      );
```
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply test/views/pages/map/static_map_test.dart]: Fixed with [98cb798](https://github.com/ProximaEPFL/proxima/pull/231/commits/98cb7981cfd137fe31c2b0c4b564357675079858)

File: test/mocks/data/map_pin.dart [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+import "package:proxima/utils/ui/map_pin.dart";
+
+//mock data to test the MapPin viewmodel
+List<MapPin> testPins = [
+  MapPin(id: "1", position: const LatLng(0, 0), callbackFunction: () {}),
+  MapPin(id: "2", position: const LatLng(1, 1), callbackFunction: () {}),
+  MapPin(id: "3", position: const LatLng(2, 2), callbackFunction: () {}),
+  MapPin(id: "4", position: const LatLng(3, 3), callbackFunction: () {}),
+  MapPin(id: "5", position: const LatLng(4, 4), callbackFunction: () {}),
+];
```
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply test/mocks/data/map_pin.dart]: Great idea! I implemented this with [2011c18](https://github.com/ProximaEPFL/proxima/pull/231/commits/2011c183ceb3bd7d05f01855e4e3f9ea04e14f2e)

File: lib/viewmodels/map_pin_view_model.dart [hunk 10]
```diff
@@ -0,0 +1,7 @@
+import "package:hooks_riverpod/hooks_riverpod.dart";
+import "package:proxima/utils/ui/map_pin.dart";
+
+// This provider is used to store the list of map pins that are displayed in the map page.
+final mapPinProvider = Provider<List<MapPin>>((ref) {
+  return List.empty();
+});
```
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply lib/viewmodels/map_pin_view_model.dart]: You are right! The pins will change with user interaction so this is a better solution. Fixed with [0511daa](https://github.com/ProximaEPFL/proxima/pull/231/commits/0511daa09362e6ceb9c2b2527b0d21f6e4228fd5)

File: lib/utils/ui/map_pin.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply lib/utils/ui/map_pin.dart]: After discussion, I think that you're rigth! Fixed with [5bc50fb](https://github.com/ProximaEPFL/proxima/pull/231/commits/5bc50fbcc4322cbf73e0b061332e49f75d633876)

File: lib/services/geolocation_service.dart [hunk 11]
```diff
@@ -18,6 +18,8 @@ class GeoLocationService {
   /// Source : https://pub.dev/documentation/geolocator_android/latest/geolocator_android/LocationAccuracy.html
   final LocationSettings locationSettings = AndroidSettings(
     accuracy: LocationAccuracy.best,
+    // We set the distance filter to 20m to avoid unnecessary updates for small movements.
+    distanceFilter: 20,
```
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply lib/services/geolocation_service.dart]: Fixed with [0022105](https://github.com/ProximaEPFL/proxima/pull/231/commits/002210522f4b17c061a9731394b68a2e8b8adf87)

File: lib/views/home_content/map/post_map.dart [hunk 12]
```diff
@@ -33,6 +39,27 @@ class PostMap extends ConsumerWidget {
...
+              ),
+            )
+            .toSet();
+      },
+      error: (error, _) {
+        throw Exception("Map pins error: $error");
+      },
+      loading: () => (),
+    );
+
```
Comment (yoannLafore) [inline lib/views/home_content/map/post_map.dart:62]: I was wondering if we could use a `CircularValue` here to manage a possible error.
I am not entirely sure however that this is the behavior we want and we may want to ask other members of the team, what do you think ?

After a little testing I think that with a `CircularValue` the code would look like:

```dart
    return CircularValue(
      value: mapPinsAsync,
      builder: (context, mapPins) {
        //list of Marker displayed on the map
        final markers = mapPins
            .map(
              (pin) => Marker(
                markerId: pin.id,
                position: pin.position,
                onTap: pin.callbackFunction,
              ),
            )
            .toSet();

        return Expanded(
          child: GoogleMap(
            key: postMapKey,
            mapType: MapType.normal,
            myLocationButtonEnabled: true,
            myLocationEnabled: true,
            zoomGesturesEnabled: true,
            zoomControlsEnabled: true,
            scrollGesturesEnabled: true,
            rotateGesturesEnabled: false,
            tiltGesturesEnabled: false,
            initialCameraPosition: CameraPosition(
              target: mapInfo.initialLocation,
              zoom: initialZoomLevel,
            ),
            circles: mapNotifier.circles,
            markers: markers,
            onMapCreated: mapNotifier.onMapCreated,
          ),
        );
      },
    );
```

File: test/mocks/data/map_pin.dart [hunk 13]
```diff
@@ -0,0 +1,18 @@
+import "package:google_maps_flutter/google_maps_flutter.dart";
+import "package:proxima/models/ui/map_pin.dart";
+
+class MapPinGenerator {
+  //generate a list of MapPin
+  List<MapPin> generateMapPins(int numberOfPins) {
```
Comment (yoannLafore) [inline test/mocks/data/map_pin.dart:6]: Nice work implementing the generator !
I just think that we can make this method static :+1:

File: lib/views/home_content/map/post_map.dart
Comment (yoannLafore) [inline lib/views/home_content/map/post_map.dart:1]: I noticed that we are throwing an expection when `positionValue` has an error.
I am not sure that this is exactly the behavior we want.
I wonder if we shouldn't display an error popup instead ?

However, I haven't found a clean way to do it with the current architecture thus if we want that I think that we could add it to the backlog.
What do you think ?

**Review by yoannLafore — COMMENTED** [review]
Thank you for bringing the modifications !
I think you did a great job with this PR :+1: 

I have just a few little suggestions below and then I believe that the PR should be ready to be merged after that.

File: test/mocks/data/map_pin.dart (hunk 13 above)
Comment (TheTexanCodeur) [reply test/mocks/data/map_pin.dart]: Fixed with [7e338be](https://github.com/ProximaEPFL/proxima/pull/231/commits/7e338bed5c379b7d804b3ce087536e3634ffe632)

File: lib/views/home_content/map/post_map.dart [hunk 14]
```diff
@@ -19,8 +20,13 @@ class PostMap extends ConsumerWidget {
 
   @override
   Widget build(BuildContext context, WidgetRef ref) {
+    // This provider is used to get information about the map.
     final mapNotifier = ref.watch(mapProvider.notifier);
 
+    // This provider is used to get the list of map pins.
+    final mapPins = ref.watch(mapPinProvider);
```
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply lib/views/home_content/map/post_map.dart]: Fixed with [a8ec2b1](https://github.com/ProximaEPFL/proxima/pull/231/commits/a8ec2b18c7c9dc73eaf0d99c4fa20476f8346de0)

File: lib/services/geolocation_service.dart [hunk 15]
```diff
@@ -4,6 +4,8 @@ import "package:cloud_firestore/cloud_firestore.dart";
 import "package:geolocator/geolocator.dart";
 import "package:hooks_riverpod/hooks_riverpod.dart";
 
+const minimalDistanceForUpdate = 5;
+
 class GeoLocationService {
```
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply lib/services/geolocation_service.dart]: Fixed with [ad758b6](https://github.com/ProximaEPFL/proxima/pull/231/commits/ad758b60cd64d5c29a3ef4e3b50c7abfffb2eab4)

File: lib/views/home_content/map/post_map.dart (hunk 12 above)
Comment (TheTexanCodeur) [reply lib/views/home_content/map/post_map.dart]: Fixed with [65fcf16](https://github.com/ProximaEPFL/proxima/pull/231/commits/65fcf1659fde502d85d00cd3998acf4720de4054)

File: lib/views/home_content/map/post_map.dart
Comment (TheTexanCodeur) [reply lib/views/home_content/map/post_map.dart:1]: With [65fcf16](https://github.com/ProximaEPFL/proxima/pull/231/commits/65fcf1659fde502d85d00cd3998acf4720de4054) an error popup is now displayed when an error occurs with the love location.

**Review by yoannLafore — APPROVED** [review]
The PR looks good to me now :+1: 

Regarding the error management, I think this is already a good improvement to display an error popup instead of throwing an exception.

Thank you for your work,
Great job :tada:

**Pull request merged**